src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.TagNode.cs

```diff
@@ -2,49 +2,36 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
 using Microsoft.VisualStudio.Text;
 using Microsoft.VisualStudio.Text.Tagging;
 
 namespace Microsoft.CodeAnalysis.Editor.Shared.Tagging;
 
 internal partial class TagSpanIntervalTree<TTag>
 {
-    private class TagNode(ITagSpan<TTag> ts, SpanTrackingMode trackingMode)
+    private readonly struct TagNode(ITagSpan<TTag> tagSpan, SpanTrackingMode trackingMode)
     {
-        public readonly TTag Tag = ts.Tag;
-        public readonly ITrackingSpan Span = ts.Span.CreateTrackingSpan(trackingMode);
```

Comment:
note: i intentionally removed the up front allocation of the tracking span.  The tracking is done on demand, only for the short period of times where we have a tag and want to map it forward, and haven't produced the updated the new tags yet.  I tried typing as well, and didn't see any of the tracking even showing up.  This is likely because we only ever have a short distance between the original text snapshot and the latest one.  So tracking from the original to the latest is extremely fast and simple without hte need for a dedicated allocated tracking span.

Reply:
i also didn't move to a model where we compute and cache this on demand because i wanted to make this a struct, and having this be a mutable struct was non-trivial in terms of ensuring it could mutate through all the interfaces and dispatch pattern we work with.